### PR TITLE
Fix the HttpRouter problem with overlapping in trail and the middle of routes

### LINKED ISSUE #378: router urls with "?"
First and foremost thanks for the fun library! Makes my integration tests pretty painless =)

Perhaps I've missed something in the docs, but:

Going from version 1.4.5 -> 1.4.6, it seems Swifter is unable to correctly parse the following url when my app requests it:

`/invoices/?page=1`

The package IS able to find the file with the json in it, but when my app makes a request to this url in `1.4.6`, I get a 404. 

The 1.4.5 version correctly finds it!

What do you advise?

All other types of routes without the `?` work fine. 

Thanks =)

### PR DESCRIPTION
**Pull request by Vkt0r**
This PR should fix #378, #381 and #376 and can be resumed in the following steps:

* Remove the set of the empty string for not matched parameters, by default should be `nil`
* Add unit tests to verify the trail dynamic identifier  as the end of the route and with more dynamic path after the end of the route and the overlapping during the middle of the route

The `findHandler` method was refactored to fix two main issues occurring during the search of a route. 

1. Nodes ending in dynamic trails were not matched correctly if they have more routes after the trail. Examples of routes are defined in the new test case `testHttpRouterShouldHandleOverlappingRoutesInTrail` added.
2. After the search traverse an incorrect path in a dynamic route the recursion was not handling the rest of the case properly regarding the `IndexingIterator<[String]` is mutated after call `next()`.

The method was refactored to use a default array with indexes instead of the `IndexingIterator<[String]` to be able to traverse the routes and search correctly and removed the part of the code causing the routes with a dynamic trail were not matched.

### CODE REVIEW

**Comment by Vkt0r** [comment]
I tried this fix in a big suite of UI Tests with a lot of overlapping in the routes as it's working as expected. Also, we have positive answers about the fix of the issue in #381 and #376. So this PR is ready to be reviewed 🚀.

File: Sources/HttpRouter.swift [hunk 1]
```diff
@@ -106,17 +109,17 @@ open class HttpRouter {
     ///   - matchedNodes: An array with the nodes matching the route
     ///   - index: The index of current position in the generator
     ///   - count: The number of elements if the route to match
-    private func findHandler(_ node: inout Node, params: inout [String: String], generator: inout IndexingIterator<[String]>, matchedNodes: inout [Node], index: Int, count: Int) {
+    private func findHandler(_ node: inout Node, params: inout [String: String], pattern: [String], matchedNodes: inout [Node], index: Int, count: Int) {
```
Comment (adamkaplan) [inline Sources/HttpRouter.swift:112]: The swift doc on this method still references generator and indexing iterator

**Review by adamkaplan — APPROVED** [review]

**Comment by adamkaplan** [comment]
Merge if you like, just one minor comment about the docs, otherwise looks good

File: Sources/HttpRouter.swift (hunk 1 above)
Comment (Vkt0r) [reply Sources/HttpRouter.swift:112]: Uhmm I forgot to update it. Let's fix it 👍.

**Comment by Vkt0r** [comment]
Thanks, @adamkaplan and welcome back 😀.

**Pull request merged**